apps/dcos_net/src/dcos_net_mesos_listener.erl

```diff
@@ -925,13 +960,88 @@ notify(TaskId, Task, #state{subs = Subs, timeout = Timeout} = State) ->
                 exit(Pid, {?MODULE, timeout}),
                 St
             end
-        end, State, Subs)
+        end, State, Subscribers)
     after
         prometheus_summary:observe(
             mesos_listener, pubsub_duration_seconds,
             [], erlang:monotonic_time() - Begin)
+    end;
+pubsub(_Message, _Subscribers, State) ->
+    State.
+
+-spec(pubsub_tasks(state()) -> state()).
+pubsub_tasks(#state{subs = Subscribers} = State) ->
+    pubsub_tasks(Subscribers, State).
+
+-spec(pubsub_tasks(#{pid() => reference()}, state()) -> state()).
+pubsub_tasks(Subscribers, #state{tasks = T, waiting_tasks = TW} = State) ->
+    Tasks = maps:without(maps:keys(TW), T),
+    pubsub({tasks, Tasks}, Subscribers, State).
+
+-spec(pubsub_task_update(task_id(), task(), state()) -> state()).
+pubsub_task_update(TaskId, Task, State) ->
+    pubsub({task_updated, TaskId, Task}, State).
+
+-spec(start_agents_readiness_timer() -> reference()).
+start_agents_readiness_timer() ->
+    Timeout = application:get_env(dcos_net,
+        mesos_agents_readiness_timeout, timer:minutes(10)),
+    erlang:start_timer(Timeout, self(), await_agents).
+
+-spec(start_tasks_readiness_timer() -> reference()).
+start_tasks_readiness_timer() ->
+    Timeout = application:get_env(dcos_net,
+        mesos_tasks_readiness_timeout, timer:minutes(1)),
```

Comment:
I have one more question here: is it possible to check how long mesos is running, so if dcos-net / mesos listener restarted and there are no `recovered_agents` in `GET_STATE` then we can safely skip this timeout?

Reply:
Interesting idea. Yes, it is possible. There is the `GET_MASTER` call:

```
  // Contains the master's information.
  message GetMaster {
    optional MasterInfo master_info = 1;

    optional double start_time = 2;

    optional double elected_time = 3;
  }
```

Do you want me to implement it in this PR or a separate one?

Reply:
It'd be a bit sad if we have to add one more call for that. Could you please check if it is not possible to re-use agents/frameworks `(re-)registered_time` from `GET_STATE`?

Reply:
Once the timeout has been changed to 10 seconds, I think it's not a big deal. If it's easy to implement, let's do it. Otherwise, it doesn't worth the effort and I'm fine with the timeout.